File: src/main/java/org/processmining/models/simple_behavior/SimpleTransition.java
package org.processmining.models.simple_behavior;

import org.deckfour.xes.model.XTrace;
import org.processmining.models.MovementResult;
import org.processmining.models.abstract_net_representation.Place;
import org.processmining.models.abstract_net_representation.Token;
import org.processmining.models.base_implementation.BaseTransition;
import org.processmining.models.descriptions.GenerationDescriptionWithNoise;
import org.processmining.models.descriptions.SimpleGenerationDescription;
import org.processmining.models.descriptions.TimeDrivenGenerationDescription;
import org.processmining.models.graphbased.directed.petrinet.elements.Transition;
import org.processmining.utils.LoggingSingleton;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Ivan Shugurov on 30.10.2014.
 */
public class SimpleTransition extends BaseTransition
{
    protected SimpleTransition(Transition transition, SimpleGenerationDescription generationDescription, Place<Token>[] inputPlaces, Place<Token>[] outputPlaces)
    {
        super(transition, generationDescription, inputPlaces, outputPlaces);
    }

    @Override
    public MovementResult move(XTrace trace)
    {
        consumeTokens();

        if (shouldDistortEvent())
        {
            SimpleGenerationDescription description = getGenerationDescription();
            GenerationDescriptionWithNoise.NoiseDescription noiseDescription = description.getNoiseDescription();

            if (noiseDescription.isSkippingTransitions())
            {
                if (noiseDescription.isUsingInternalTransitions() || noiseDescription.isUsingExternalTransitions())
                {
                    boolean skipEvent = random.nextBoolean();

                    if (skipEvent)
                    {
                        //ignore this case in order to skip the event
                    }
                    else
                    {
                        logNoiseAndTransition(trace);
                    }
                }
                else
                {
                    //ignore this case in order to skip the event
                }
            }
            else
            {
                logNoiseAndTransition(trace);
            }
        }
        else
        {
            logTransition(trace);
        }

        produceTokens();
        return new MovementResult();
    }

    protected void logNoiseAndTransition(XTrace trace)
    {
        boolean noiseFirst = random.nextBoolean();

        if (noiseFirst)
        {
            logNoiseEvent(trace);
            logTransition(trace);
        }
        else
        {
            LoggingSingleton.log(trace, getNode());
            logTransition(trace);
        }
    }

    private void logTransition(XTrace trace)
    {
        Transition realTransition = (Transition) getNode();
        if (!realTransition.isInvisible())
        {
            LoggingSingleton.log(trace, getNode());
        }
    }

    protected void logNoiseEvent(XTrace trace)
    {
        List<Object> noiseEvents = getNoiseEventsBasedOnSettings();
        if (!noiseEvents.isEmpty())
        {
            int index = random.nextInt(noiseEvents.size());
            Object event = noiseEvents.get(index);
            LoggingSingleton.log(trace, event);
        }
    }

    private List<Object> getNoiseEventsBasedOnSettings()
    {
        List<Object> noiseEvents = new ArrayList<Object>();
        GenerationDescriptionWithNoise.NoiseDescription noiseDescription = getGenerationDescription().getNoiseDescription();
        if (noiseDescription.isUsingInternalTransitions())
        {
            noiseEvents.addAll(noiseDescription.getInternalTransitions());   //TODO почему здесь и в версии с временами по-разному работает этот метод?
        }
        if (noiseDescription.isUsingExternalTransitions())
        {
            noiseEvents.addAll(noiseDescription.getArtificialNoiseEvents());
        }
        return noiseEvents;
    }

    private boolean shouldDistortEvent()
    {
        GenerationDescriptionWithNoise description = getGenerationDescription();
        if (description.isUsingNoise())
        {
            GenerationDescriptionWithNoise.NoiseDescription noiseDescription = description.getNoiseDescription();
            if (noiseDescription.getNoisedLevel() >= random.nextInt(GenerationDescriptionWithNoise.NoiseDescription.MAX_NOISE_LEVEL + 1))  //use noise transitions
            {
                return true;
            }
        }
        return false;
    }

    @Override
    public SimpleGenerationDescription getGenerationDescription()
    {
        return (SimpleGenerationDescription) super.getGenerationDescription();
    }

    public static class SimpleTransitionBuilder
    {
        private final Transition transition;
        private final SimpleGenerationDescription description;
        private final List<Place<Token>> inputPlaces = new ArrayList<Place<Token>>();
        private final List<Place<Token>> outputPlaces = new ArrayList<Place<Token>>();


        public SimpleTransitionBuilder(Transition transition, SimpleGenerationDescription description)
        {
            if (transition == null)
            {
                throw new NullPointerException("Transition cannot be equal to null");
            }

            if (description == null)
            {
                throw new NullPointerException("Generation description cannot be null");
            }

            this.transition = transition;
            this.description = description;
        }

        public SimpleTransitionBuilder inputPlace(Place<Token> inputPlace)
        {
            if (inputPlace == null)
            {
                throw new NullPointerException("Place cannot be equal to null");
            }

            inputPlaces.add(inputPlace);

            return this;
        }

        public SimpleTransitionBuilder outputPlace(Place<Token> outputPlace)
        {
            if (outputPlace == null)
            {
                throw new NullPointerException("Place cannot be equal to null");
            }

            outputPlaces.add(outputPlace);

            return this;
        }

        public SimpleTransition build()
        {
            @SuppressWarnings("unchecked")
            Place<Token>[] inputPlacesArray = inputPlaces.toArray(new Place[inputPlaces.size()]);
            @SuppressWarnings("unchecked")
            Place<Token>[] outputPlacesArray = outputPlaces.toArray(new Place[outputPlaces.size()]);

            return new SimpleTransition(transition, description, inputPlacesArray, outputPlacesArray);
        }
    }
}

File: console/src/main/kotlin/org/processmining/models/simple_behavior/SimpleTransition.kt
package org.processmining.models.simple_behavior

import org.deckfour.xes.model.XTrace
import org.processmining.models.GenerationDescription
import org.processmining.models.MovementResult
import org.processmining.models.abstract_net_representation.Place
import org.processmining.models.abstract_net_representation.Token
import org.processmining.models.abstract_net_representation.WeightedPlace
import org.processmining.models.base_implementation.BaseTransition
import org.processmining.models.descriptions.GenerationDescriptionWithNoise
import org.processmining.models.descriptions.SimpleGenerationDescription
import org.processmining.utils.LoggingSingleton
import java.util.*
import kotlin.random.Random

private typealias WPlace = WeightedPlace<Token, Place<Token>>
/**
 * Created by Ivan Shugurov on 30.10.2014.
 */
class SimpleTransition(
        node: org.processmining.models.graphbased.directed.petrinet.elements.Transition,
        generationDescription: GenerationDescription,
        inputPlaces: List<WPlace>,
        outputPlaces: List<WPlace>,
        inputInhibitorArcPlaces: List<Place<Token>> = listOf(),
        inputResetArcPlaces: List<Place<Token>> = listOf()
) : BaseTransition(node, generationDescription, inputPlaces, outputPlaces, inputInhibitorArcPlaces, inputResetArcPlaces) {
    
    //TODO почему здесь и в версии с временами по-разному работает этот метод?
    private val noiseEventsBasedOnSettings: List<Any>
        get() {
            val noiseEvents = ArrayList<Any>()
            val noiseDescription = generationDescription.noiseDescription
            
            if (noiseDescription.isUsingInternalTransitions) {
                noiseEvents.addAll(noiseDescription.internalTransitions)
            }
            if (noiseDescription.isUsingExternalTransitions) {
                noiseEvents.addAll(noiseDescription.artificialNoiseEvents)
            }
            return noiseEvents
        }
    
    override fun move(trace: XTrace): MovementResult<*>? {
        consumeTokens()
        
        if (shouldDistortEvent()) {
            val description = generationDescription
            val noiseDescription = description.noiseDescription
            
            if (noiseDescription.isSkippingTransitions) {
                if (noiseDescription.isUsingInternalTransitions || noiseDescription.isUsingExternalTransitions) {
                    val skipEvent = Random.nextBoolean()
                    
                    if (skipEvent) {
                        //ignore this case in order to skip the event
                    } else {
                        logNoiseAndTransition(trace)
                    }
                } else {
                    //ignore this case in order to skip the event
                    // (don't use any of external/internal transitions, so just skip)
                }
            } else {
                logNoiseAndTransition(trace)
            }
        } else {
            logTransition(trace)
        }
        
        produceTokens()
        return MovementResult<Token>()
    }
    
    protected fun logNoiseAndTransition(trace: XTrace) {
        val noiseFirst = Random.nextBoolean()
        
        if (noiseFirst) {
            logNoiseEvent(trace)
            logTransition(trace)
        } else {
            // duplicate transition
            LoggingSingleton.log(trace, node)
            logTransition(trace)
        }
    }
    
    private fun logTransition(trace: XTrace) {
        val realTransition = node
        if (!realTransition.isInvisible) {
            LoggingSingleton.log(trace, node)
        }
    }
    
    protected fun logNoiseEvent(trace: XTrace) {
        val noiseEvents = noiseEventsBasedOnSettings
        if (!noiseEvents.isEmpty()) {
            val index = Random.nextInt(noiseEvents.size)
            val event = noiseEvents[index]
            LoggingSingleton.log(trace, event)
        }
    }
    
    /** @return true if we should use noise events */
    private fun shouldDistortEvent(): Boolean {
        val description = generationDescription
        if (description.isUsingNoise) {
            val noiseDescription = description.noiseDescription
            if (noiseDescription.noisedLevel >= Random.nextInt(GenerationDescriptionWithNoise.MAX_NOISE_LEVEL + 1)) {
                //use noise transitions
                return true
            }
        }
        return false
    }
    
    override val generationDescription: SimpleGenerationDescription
        get() = super.generationDescription as SimpleGenerationDescription
}
